Invalidate type cache on schema changes affecting statement result.

vitaly-burovoy opened the pull request:
Almost all explanation is written in the commit message.

The rest I want to mention is at least one feature is left untouched. It is the `BaseCursor` where the `bind_execute` and `bind` calls should be wrapped by the same way the `bind_execute` in the `PreparedStatement.__bind_execute` was, but I'd leave it for the other commit (may be by the other author).

1st1 commented on asyncpg/connection.py:
```diff
@@ -1222,7 +1222,10 @@ def _on_release(self, stacklevel=1):
         self._stmt_cache.clear()
 
-    def _drop_global_statement_cache(self):
+    def _drop_global_statement_cache(self, drop_type_cache=False):
```
Make this a keyword-only parameter.

elprans replied:
Arguably, type cache and statement cache are separate concerns.  I wouldn't lump both of them into a single method.

vitaly-burovoy replied:
> are separate concerns

Yes, but it has no sense to clear type cache without statement one because statement cache keeps prepared statements which has their own local copy of codecs.

elprans replied:
It's more nuanced than that.  We are only concerning ourselves with type cache because of composite types, which may change their shape.  We are dealing with two distinct types of failures induced by stale cache: 1) A `RevalidateCachedQuery` failure due to the change of the interpretation of the result type (which may, or may not be a result of an explicit DDL action). 2) Stale composite codec configuration due to an explicit `ALTER TABLE` or `ALTER TYPE`.  IMO, both of these cases deserve separate methods rather than a toggle via a boolean kwarg.

> Yes, but it has no sense to clear type cache without statement one because statement cache keeps prepared statements which have their own local copy of codecs.

Then call `_drop_statement_cache` from `_drop_codec_cache`.

1st1 commented on asyncpg/connection.py:
```diff
@@ -1272,6 +1275,17 @@ def _drop_global_statement_cache(self):
                     after = time.monotonic()
                     timeout -= after - before
 
+        except exceptions.InternalClientError:
```
I'd like to see a specific exception being used here, like `OutdatedTypeCacheError`.  It can be derived from `InternalClientError` to make it easier to write code that just want to handle a broad set of errors.

1st1 commented on asyncpg/connection.py:
```diff
@@ -1510,7 +1524,8 @@ def get(self, query, *, promote=True):
         if entry._statement.closed:
             # Happens in unittests when we call `stmt._state.mark_closed()`
-            # manually.
+            # manually and when a prepared statement closes itself on type
```
and -> or

1st1 commented on asyncpg/prepared_stmt.py:
```diff
@@ -198,8 +198,17 @@ def cursor(self, *args, prefetch=None,
+        try:
+            data, status, _ = await protocol.bind_execute(
+                self._state, args, '', limit, True, timeout)
+        except exceptions.InternalClientError:
```
Same thing here—use a specific exception.

1st1 commented on asyncpg/protocol/codecs/base.pyx:
```diff
@@ -191,15 +194,28 @@ cdef class Codec:
 
         elem_count = <ssize_t><uint32_t>hton.unpack_int32(buf.read(4))
         result = record.ApgRecord_New(self.element_names, elem_count)
+        if elem_count != len(self.element_type_oids):
```
Move this one line up, before we create a `Record` object.

1st1 commented:
> I'm wondering if we should add a public cache busting method, so that instead of catching the exception and retrying, one would clear the cache pre-emptively:

-1.

A more future proof solution is to suggest to simply reconnect (create a new connection/pool) in these cases.  It's not a normal situation when a program alters tables/types.

elprans commented:
DML followed by data queries in a single transaction is a valid pattern in migrations.

vitaly-burovoy commented:
> I'm wondering if we should add a public cache busting method

I do not like when a lot of logically different things are placed in a single commit.

I planned it as a separated PR because it deserves a different comment in a commit message, but yes, it would have just a renaming (or creating an alias) `_drop_global_statement_cache` => `drop_statement_cache` or similar one.

Primarily I see such method for being called from NOTIFY messages handlers which are sent from migration tools (sometimes that tools are alone scripts/services) or even from event triggers:
https://www.postgresql.org/docs/devel/static/event-trigger-definition.html

vitaly-burovoy commented:
> A more future proof solution is to suggest to simply reconnect (create a new connection/pool) in these cases.

Really? If it is so easy (just reconnect after changing schema), why do we have pool's behaviour tests in the `asyncpg/tests/test_cache_invalidation.py`?

In fact the type cache is shared across `DataCodecConfig`-s via the `ConnectionSettings` by the "`cache_key`" which is just a tuple `(address, database)`. 

Proofs:
```asyncpg/protocol/protocol.pyx:102```
```asyncpg/protocol/settings.pyx:16```
```asyncpg/protocol/codecs/base.pyx:378```

elprans commented on asyncpg/connection.py:
```diff
@@ -1222,7 +1222,10 @@ def _on_release(self, stacklevel=1):
+    def _drop_global_statement_cache(self, drop_type_cache=False):
+        if drop_type_cache:
+            # It is enough to clear the type cache only once.
+            self._protocol.get_settings().clear_type_cache()
```
Type cache is global across connection to _all_ databases. Please make the invalidation database-specific.

vitaly-burovoy replied:
Am I mistaken or `self._protocol` is database-specific connection?
```asyncpg/connect_utils.py:261```

vitaly-burovoy replied:
Also:
```asyncpg/protocol/protocol.pyx:102```
```asyncpg/protocol/protocol.pyx:131```

vitaly-burovoy commented:
> so that instead of catching the exception and retrying

Yes, it is a good solution, but we still have to support catching/raising these exceptions because we can not expect this method is called or all manually prepared statements are invalidated by a user.

The other case is currently running queries which start between DDL is done at the server's side but before we clear the cache.

1st1 commented:
> Really? If it is so easy (just reconnect after changing schema), why do we have pool's behaviour tests in the

I suggest you to calm down ;)

1st1 commented:
> DML followed by data queries in a single transaction is a valid pattern in migrations.

Right, that's a valid use case. As we discussed off list, let's add `Connection.refresh()` and `Pool.refresh()` coroutine methods to reset statement and type caches.

vitaly-burovoy commented:
> let's add `Connection.refresh()`

Naming is a big deal.

I think that name gives nothing about what it does (refreshes what?) and people who will use it will know for sure what they will going to do.

I would leave object ("cache") under the action ("refresh"|"drop"|"clear") as I mentioned in https://github.com/MagicStack/asyncpg/pull/235#issuecomment-349469079

Moreover it does not re-`fresh`. I understand this word as clear and fetch again immediately or fetch and update changed items.

vitaly-burovoy commented:
In the new branch:
* `drop_type_cache` is a keyword-only parameter;
* a new `OutdatedTypeCacheError` is used;
* "and -> or" is done;
* "before we create a Record object" good catch, I overlooked it;
* some rewordings (comments and the commit message).

Left untouched:
* "Please make the invalidation database-specific.": It is already database-specific;
* "add `Connection.refresh()`": I'd like a different name and a separate commit.

elprans commented:
> Left untouched:
> * "Please make the invalidation database-specific.": It is already database-specific;

It's not.  `TYPE_CODECS_CACHE` is a global dict keyed by `(addr, database)`

vitaly-burovoy commented:
> `TYPE_CODECS_CACHE` is a global dict

Correct, but the `clear_type_cache` affects not whole `TYPE_CODECS_CACHE`, but `TYPE_CODECS_CACHE[(conn_addr, conn_database)]`. Does not it?

elprans commented:
Yes. Just wanted to make sure it stays that way.

elprans commented:
> Naming is a big deal.
> I think that name gives nothing about what it does (refreshes what?) and people who will use it will know for sure what they will going to do.

`refresh` may not be the best name, but I don't like the `drop|clear_statement_cache` name either.  It says too much about the implementation details, and is not very future-proof.  We may need to do more than drop the statement cache in the future.  The idea behind a name like `refresh` is to be a generic instruction to bring the connection into a "fresh" state and dismiss all assumptions about the database schema post-DDL.

`Connection.reset()` is not called `unlock_close_cursors_unlisten_reset()` for the same reason.

1st1 commented on asyncpg/protocol/codecs/base.pyx:
```diff
@@ -190,16 +193,29 @@ cdef class Codec:
         elem_count = <ssize_t><uint32_t>hton.unpack_int32(buf.read(4))
+        if elem_count != len(self.element_type_oids):
+            raise OutdatedTypeCacheError(
+                'unexpected number of attributes of composite type {!r}.{!r}: '
```
You're using `{!r}` for `self.schema` and `self.name`.  Wouldn't it render them in quotes, like this "... of composite type 'abc'.'def': ..."?

vitaly-burovoy replied:
Right. It is intended to avoid ambiguity when either `schema` or `name` has dots. It is similar to PG's `format('%I.%I', schema, name)` but I don't mind quotes type (single or double) and their presence in some cases.

It is just convenient format for parsing.

elprans replied:
Please don't rely on parsing when you can avoid it.  Store the data as attributes on the exception.  Postgres' native exceptions have the `schema` and `data_type_name` attributes for similar purposes.

1st1 replied:
Here's a list of attribute names that asyncpg errors can have.  I'd try to use only them.

https://github.com/MagicStack/asyncpg/blob/59e2878bef959cfab35ebd19c211477c1daf1101/asyncpg/exceptions/_base.py#L26-L45

vitaly-burovoy replied:
Great! I almost forgot it.
I hope the `OutdatedTypeCacheError` should neither be inherited from the `PostgresMessage` class nor have the `PostgresMessageMeta` as a metaclass.

In the other case we would face troubles of filling the `severity` and `sqlstate` fields which are required.

elprans replied:
It should not inherit from `PostgresMessage`, since it does not originate from the server.  But it still makes sense to keep it close in shape to server errors.  With respect to the attributes, let's stick to just `schema` and `data_type_name` (that's what you were intending to parse out of the message anyway).

1st1 replied:
@elprans you mean `schema_name`?

elprans replied:
Yes, `schema_name`.

vitaly-burovoy replied:
@elprans, thank you! I just wanted be sure I understand @1st1's hint correctly.
I'm sorry, I can push the discussed improvements/fixes only at the weekend.

vitaly-burovoy commented:
> `Connection.reset()` is not called `unlock_close_cursors_unlisten_reset()` for the same reason.

:-)

Well... `reset` is a known word in DB's context¹, whereas `refresh` is a new term there. I agree no sense for the `asyncpg`'s users to know its internals like "statement" and "type" caches, but I think those who will use it will know that they want to clear exactly the cache (one or the other or both).

May be we need to define a function named "`Connection.clear_cache()`" which affects both caches?

>  We may need to do more than drop the statement cache in the future.

Since we can not imagine what we will need in the future we can now create a function with minimum info in its name (like mentioned above `clear_cache`) and then (in the future) create a new one with a different name reflected a necessary action inside and mark the current as deprecated.

¹ https://www.postgresql.org/docs/current/static/sql-reset.html

elprans commented:
> I think those who will use it will know that they want to clear exactly the cache (one or the other or both)

We are dealing with an unfortunate case of a leaky abstraction here.  The goal is _not_ to encourage the users to tinker with connection internals (and doing all kinds of foot-shooting in the process).  The goal is to handle the very specific case of "the database schema has changed, let the driver know".  The fact that there are various caches that need clearing is incidental and internal to the driver.

>  Then (in the future) create a new one with a different name reflected a necessary action inside and mark the current as deprecated.

This is exactly what I would like to avoid here. Deprecations aren't cheap.

vitaly-burovoy commented:
> The fact that there are various caches that need clearing is incidental and internal to the driver.

I offered a name `clear_cache()` where there is nothing about internals except the fact it works with a Pool's/ Connection's caching system (no matter how many of them internally), and I don't think people will use it "just for fun" without understanding why they have to use it.

Really, I do not insist on giving exactly that name. I just want to mention your version of a name gives no information what happens there and for what it is introduced in the `Connection`/`Pool` classes.

1st1 commented on asyncpg/exceptions/_base.py:
```diff
@@ -221,6 +221,10 @@ class InternalClientError(Exception):
 
 
+class OutdatedTypeCacheError(InternalClientError):
+    pass
```
Need a docstring for this class.

vitaly-burovoy commented:
The next round...

>> I'm wondering if we should add a public cache busting method

> it deserves a different comment in a commit message

Moved to a separate commit (+usage example) and wiped out `drop_type_cache` parameter. We have not finished a discussion about a name of the method, I set it as [`drop_cache()`](https://github.com/MagicStack/asyncpg/pull/235#issuecomment-349791951), but still open for arguments.

> Yes. Just wanted to make sure it stays that way.

A test added as [promised](https://github.com/MagicStack/asyncpg/pull/235#issuecomment-349710434).

> Need a docstring for this class.

Added.

> The latest version looks OK, except one nit about using {!r} for strings formatting.

Moved to the `OutdatedTypeCacheError` attributes as [discussed](https://github.com/MagicStack/asyncpg/pull/235#discussion_r155418393).

Plus changes to examples (termination line should be added¹ when function's body finishes) and a typo fix in a help string.

¹ https://docs.python.org/3/tutorial/introduction.html#first-steps-towards-programming

vitaly-burovoy commented:
> would be useful beyond the case in this PR

As of now I can't imagine what's for it would be useful. I think a name of any method should reflect its purpose. The `reset` method is "set settings to default" from the SQL standard, whereas the `refresh` name gives me nothing about its purpose. I [said](https://github.com/MagicStack/asyncpg/pull/235#issuecomment-349480167) that for me it means it _fetches_ requested data (like "refresh a WEB-page").

On the other hand I can't imagine what would we have to do more than dropping caches and it is really more or less rare case for people.


> Maybe `handle_schema_changes()` or something similar?

Yes, it looks much better than the `refresh()` one. Would a shorter `Connection.schema_changed()` be a good name?

I know that many methods have a verb in their names, but there are also `cursor()` and `transaction()` which do not.

elprans commented:
> Would a shorter `Connection.schema_changed()` be a good name? I know that many methods have a verb in their names, but there are also `cursor()` and `transaction()` which do not.

The ones that are not verbs are constructors.  `schema_changed()` reads like it returns a "changed schema".  The length in this case is not a significant concern, as the method is not supposed to be used very often.

vitaly-burovoy commented:
> reads like it returns a "changed schema".

Oops.

OK, something like `handle_schema_changes()` seems goods enough. Let me know whether to use that name or something similar to.

1st1 commented on asyncpg/connection.py:
```diff
@@ -1231,6 +1236,47 @@ def _drop_global_statement_cache(self):
         else:
             self._drop_local_statement_cache()
 
+    def drop_cache(self):
```
This method should be an `async def` method, because we want it to be future-proof.

vitaly-burovoy replied:
OK, I understand the requirement, I'll push it with a new name.

1st1 commented:
> OK, something like handle_schema_changes() seems goods enough.

I'm big -1 on both `handle_schema_changes()` (because it doesn't "handle" them, really), and on `drop_caches()`.

I'm still in favour of `refresh()`.

vitaly-burovoy commented:
The `refresh()` name is tricky.

What should a connection refresh? Transactions/cursors? Data (after a query failed)? Do some sort of ping and/or reconnect?

Outside of this PR it looses the context and gives questions being in a list of methods.

---
> because it doesn't "handle" them, really

What does it do?

elprans commented:
OK, let's take a step back. The statement cache and the type cache are fundamentally reflections of the database schema state.  DDL on the server side invalidates that state, so the driver needs to update its reflection.  Note: _"update"_, not _"invalidate"_ or _"drop"_, since the latter, again, talk about the specific internal process, which shouldn't matter to the caller.  The effect of the "update" call is that subsequent queries succeed, that's all.

Thus, I propose `reload_schema_state()` as the name.

vitaly-burovoy commented:
Since the `schema` word is in the name it looks much better. For me the `reload` means "load at that moment" especially when the method is declared as `async def`.

Therefore the `update_schema_state()` is more preferable than the first one.

1st1 commented:
`update_schema_state()` could imply that it's updating the schema in the DB, whereas in reality it's only updating some internal state of the **connection** itself. -1 on it.

`reload_schema_state()` is the best option so far, IMO.

1st1 commented:
Vitaly, let's stop the bike shedding, as everybody is getting very annoyed about this issue. `reload_schema_state()` it is.

elprans commented on asyncpg/connection.py:
```diff
@@ -1231,6 +1236,47 @@ def _drop_global_statement_cache(self):
             self._drop_local_statement_cache()
 
+    def drop_cache(self):
+        """Drop internally cached objects in the connection.
```
```python
"""Indicate that the database schema information must be reloaded.

For performance reasons, asyncpg caches certain aspects of the 
database schema, such as the layout of composite types.  Consequently,
when the database schema changes, and asyncpg is not able to
gracefully recover from an error caused by outdated schema 
assumptions, an :exc:`~asyncpg.exceptions.OutdatedSchemaCacheError` 
is raised.  To prevent the exception this method may be used to inform 
asyncpg that the database schema has changed.

Example:
...
"""
```

vitaly-burovoy commented:
Now:
1. the method is defined as `async def reload_schema_state(self)`;
2. the exception has the name `OutdatedSchemaCacheError`;
3. the docstring has changed to one [offered by @elprans](https://github.com/MagicStack/asyncpg/pull/235#discussion_r156765798).

1st1 commented on asyncpg/connection.py:
```diff
@@ -423,6 +423,7 @@ def cursor(self, query, *args, prefetch=None, timeout=None):
             ...         'mytable', columns=('foo', 'bar'),
             ...         output='file.csv', format='csv')
             ...     print(result)
+            ...
```
Let's not change any lines of code/comments/docstrings that aren't directly relevant to the PR.

vitaly-burovoy replied:
I created a separated commit for it (to not mix it with lines in the commit relevant to the PR).

Does it worth a separate PR? The git history does not track PRs (may be except ranges between GPG signatures).

1st1 commented on tests/test_cache_invalidation.py:
```diff
@@ -5,8 +5,13 @@
-import asyncpg
 from asyncpg import _testbase as tb
+from asyncpg.exceptions import InterfaceError, InvalidCachedStatementError
+from asyncpg.exceptions import OutdatedSchemaCacheError
```
We allow only module imports in `*.py` files.  Access exceptions like `asyncpg. OutdatedSchemaCacheError`

vitaly-burovoy replied:
It is a test.
Moreover the `test_connect.py` has
```python
from asyncpg.serverversion import split_server_version_string
```
(khm, ok, the above is from one of my commit, but there were no nitpickings)

the `test_record.py` has
```python
from asyncpg.protocol.protocol import _create_record as Record
```
(that one is written by you in the 487f0f17e148ba6b4bfb7f03a438bb8ba02650a1)
---
In fact the only reason why I used it that way was long rows with `self.assertRaisesRegex(…)`.

OK, I've just changed it to fit your requirements.

1st1 commented:
Thanks @vitaly-burovoy, looks good!  I've left a couple of nit comments (last ones).

vitaly-burovoy commented:
Pushed.

elprans commented:
Sorry, you seem to have squashed the wrong commits.  There should be two commits: cache invalidation and your typo fixes.

vitaly-burovoy commented:
Not a problem.

One sad thing is absence of a separate logical commit message as I [mentioned](https://github.com/MagicStack/asyncpg/pull/235#issuecomment-349469079).

Nevertheless — pushed.

elprans commented:
Merged! Thank you Vitaly!

vitaly-burovoy commented:
Thank you, Elvis and Yury, very much: your requests for changes made the patch be better.
Special appreciation for your patience during the conversation.

elprans commented:
It's a pleasure, Vitaly. Thanks for contributing!
